### DISCUSSION TITLE
Boundary conditions from SMC grids

### DISCUSSION THREAD

**Discussion by EmilioEchevarria** [NOAA-EMC/WW3#881] [Q&A] [answered]
Dear WAVEWATCH modellers, 

I'm trying to understand the logic for generating boundary conditions for regional grids when using a SMC grid, I would appreciate any advice/comment.

If I understand correctly, we define the segments of output boundary points at the end of ww3_grid.inp. For each of these locations, to compute the boundary conditions, WW3 seems to grab the 4 surrounding points and compute some interpolation weights (https://github.com/NOAA-EMC/WW3/blob/develop/model/src/w3gridmd.F90#L5755-L5819). These interpolation weights come from https://github.com/NOAA-EMC/WW3/blob/develop/model/src/w3gsrumd.F90#L2710-L2733. It seems that if **at least one** of the 4 surrounding points is "sea" then the calculation is carried out (see [here](https://github.com/NOAA-EMC/WW3/blob/develop/model/src/w3gridmd.F90#L5775-L5782), notice the `.OR.`s), and if there are points on land the interpolation weight for that point is zero and the rest get updated. 

Now, all of this makes sense to me in a regular grid scenario. What happens exactly when we use a SMC grid, where we only define an array of sea points? The land-sea mask logic seems to still be there somehow (e.g., I see the file `mask.ww3` being generated), how is this calculated for the SMC grid? I can see the mask is defined on the base resolution of the SMC grid. But for interpolating to the boundary points location, how does it work when for example we have higher resolution points close to the coast? If I understand correctly, WW3 will still grab the 4 surrounding points, but how are these calculated in the SMC case, where you have different resolutions when getting close to the coast? As far as I can see the 4 surrounding points are defined on the base resolution of the SMC grid, but what if we have points at a higher resolution than the base resolution?

I am probably missing something obvious, but if someone can point me to the relevant pieces of code for this issue, I would really appreciate it. 

Kind regards,
Emilio

**Comment by JessicaMeixner-NOAA** [comment]
@ukmo-ccbunney  or @ukmo-jianguo-li could you help answer this question?

**Comment by EmilioEchevarria** [comment]
Thanks @JessicaMeixner-NOAA for the quick reply. I only wanted to add that I now understand just a little bit more of what's going on. What WW3 writes in the files `nest#.ww3` (after the header) is: each bnd point coordinates and the indices and interpolation weights of the 4 surrounding points ([here](https://github.com/NOAA-EMC/WW3/blob/develop/model/src/w3iobcmd.F90#L376-L380)?). I'm interested in understanding how these 4 surrounding points are calculated. Looking at `w3gridmd.F90`, it looks like they're defined by IXR and IYR, which in turn come from calling W3GRMP. 
So, looking at [these lines](https://github.com/NOAA-EMC/WW3/blob/develop/model/src/w3gsrumd.F90#L2710-L2713), it seems that the 4 surrounding points are calculated with the function [W3GFCL](https://github.com/NOAA-EMC/WW3/blob/develop/model/src/w3gsrumd.F90#L1252-L1260). 

Again, in the case of a regular grid this makes to me, but I don't know how it works with a SMC grid of various levels of resolution (e.g., the 4 "surrounding" points might not all be on the base resolution of the grid, what happens in that case?). I'm running a few tests and finding some small issues with that, the wave height at the boundary segment is lower than the interpolated wave height from the gridded output, and I wouldn't expect a significant difference there. Is it possible to modify W3GFCL to treat SMC grids differently, and grab the 4 **closest** points to each boundary point? Then we could modify the interpolation weights to be proportional to the inverse distance, or similar. 

Many thanks,
Emilio

**Comment by ukmo-jianguo-li** [comment]
Hi, Emilio  cc Jessica

You are right, the boundary interpolation part is based on a regular grid model.  SMC grid has a mirror regular grid at its base resolution so the boundary spectral output from a SMC grid model will use the mirror regular grid for the interpolation.  In refined area, the mirror grid is mapped to the last SMC grid cell within the base resolution mesh.  This will usually the largest cell within the given regular grid mesh area.  Nevertheless, it is usually not at the base resolution grid centre and hence it will definitely have some errors if the output boundary spectra are from a refined area of the SMC grid.  If any of you could update the boundary interpolation subroutine to take into account of the refinement, it will be great.

Emilio, I like to know a bit more about your boundary problem.  Are the boundary spectra are from a SMC grid model and in a refined area?  If not, you may check whether other factors are causing the problem, such as the boundary point is close to coastline or the model spectra is underestimated in the first place.  Usually, wave spectra do not change much over one grid length if they are in an open water area and the interpolation would not show much difference even if the interpolation distance has some errors.

Hope this may help.
Kind regards.
Jian-Guo

**Comment by EmilioEchevarria** [comment]
Thank you very much @ukmo-jianguo-li  for your explanation, now it makes sense. I can say pretty confidently now that the issue I have with the boundaries is not because of being too close to land, but because they are in a refined area of the SMC. One edge of the child grid sits on the base resolution of the SMC, and there is no problem there.

Then the only remaining question I have is, where exactly in the WW3 source code is that mirror grid at the base resolution calculated?

**Comment by ukmo-jianguo-li** [comment]
Hi, Emilio

The base resolution regular grid is setup by the regular grid parameters inside ww3_grid.inp.  For instance, the following parameter lines specify our global SMC 3-6-12-25 km global grid.

     'SMCG'  T  'SMPL'
   1024    736
  0.35156250  0.23437500   1.
  0.17578125 -82.3828125   1.

Note the 'SMCG' grid type is a recent update in version 7 for the multi-grid option with SMC sub-grids.  If you are using old versions, the regular grid type ''RECT' should be used.  Inside the ww3_grid program, the regular grid mapping arrays are setup with SMC grid cell array.  It is possible to set up the mirror regular grid at the finest SMC grid resolution, but it will waste a lot of storage for regular grid output.

Hope this helps.
Kind regards.
Jian-Guo

**Comment by EmilioEchevarria** [comment]
Hi @ukmo-jianguo-li thank you so much, it does help a lot. 

I'm interested in that option, how could I go about trying to set up the mirror grid not at the base resolution (size-4 cells in my case), not at the highest resolution (size-1) but at the first refinement level (size-2 cells)? Could you point me to the bits of code that should be modified?

Many thanks,
Emilio

**Comment by ukmo-jianguo-li** [comment, marked as answer]
Hi, Emilio

It could be done but you need to modify the ww3_grid program first.  The mirror grid is represented by the mapping array (MAPFS), which links each regular grid with a sea point.  You just need to modify the mapping array according to your chosen resolution level.  I did not try this so could not say more than these.  The MAPSF, mapping a sea point to a regular grid point, also needs to be altered accordingly as far as I see.  There might be other parts to be modified but I could not figure out right now.  It needs a careful checking of the source code.

It is a big change and I would not say it is a piece of cake.  But it will be useful if you could make it work.

Good luck.
Jian-Guo

**Comment by EmilioEchevarria** [comment]
Thank you Dr. Li, 

That's good to know. But I believe I will try a slightly different approach, I feel like it would be easier to select the 4 nearest SMC cells and compute inverse distance interpolation weights from these, hopefully that's a bit more straightforward.

Best regards,
Emilio
